src/UserSession.php

```diff
@@ -171,4 +246,141 @@ private function tokenFromResponse(array $data): Token
 
         return $token;
     }
+
+    /**
+     * Send a request, retrying once if the authentication token is invalid.
+     *
+     * This method creates three promises. The outer promise is what is
+     * returned, and itself contains two promises. The first promise is the
+     * first attempt to make an authenticated request. If it fails, a second
+     * promise is created that explicitly rejects the outer promise, as we only
+     * want to retry once.
+     *
+     * @param string                                $method  HTTP method
+     * @param string|\Psr\Http\Message\UriInterface $uri     URI object or string.
+     * @param array                                 $options Request options to apply.
+     *
+     * @return \GuzzleHttp\Promise\PromiseInterface|\Psr\Http\Message\RequestInterface
+     */
+    private function sendAsyncWithRetry(RequestInterface $request, array $options)
+    {
+        /** @var \GuzzleHttp\Promise\Promise $promise */
+        $promise = new Promise(function () use (&$promise, $request, $options) {
```

Comment:
This was easily the most complex part of the code to figure out, as the initial implementation was actually blocking. Hopefully the doc comment above makes this code a little clearer.

The `&$promise` was new to me - it lets the anonymous function access $promise, even though it technically hasn't been defined yet when the function is being created. Neat!